### PR TITLE
Add Brainpool Curves and Expand ECDSA support in openpgp

### PR DESCRIPTION
**Pull request by 0xhaven**
To address https://github.com/keybase/client/issues/1310 (inability to import keys from Brainpool curves)

### CODE REVIEW

File: brainpool/brainpool.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	p384r1.B, _ = new(big.Int).SetString("04A8C7DD22CE28268B39B55416F0447C2FB77DE107DCD2A62E880EA53EEB62D57CB4390295DBC9943AB78696FA504C11", 16)
+	p384r1.Gx, _ = new(big.Int).SetString("1D1C64F068CF45FFA2A63A81B7C13F6B8847A3E77EF14FE3DB7FCAFE0CBD10E8E826E03436D646AAEF87B2E247D4AF1E", 16)
+	p384r1.Gy, _ = new(big.Int).SetString("8ABE1D7520F9C2A45CB1EB8E95CFD55262B70B29FEEC5864E19C054FF99129280E4646217791811142820341263C5315", 16)
+	p384r1.BitSize = 384
+}
+
+func initP512r1() {
+	p512r1 = &elliptic.CurveParams{Name: "brainpoolP512r1"}
+	p512r1.P, _ = new(big.Int).SetString("AADD9DB8DBE9C48B3FD4E6AE33C9FC07CB308DB3B3C9D20ED6639CCA703308717D4D9B009BC66842AECDA12AE6A380E62881FF2F2D82C68528AA6056583A48F3", 16)
+	p512r1.N, _ = new(big.Int).SetString("AADD9DB8DBE9C48B3FD4E6AE33C9FC07CB308DB3B3C9D20ED6639CCA70330870553E5C414CA92619418661197FAC10471DB1D381085DDADDB58796829CA90069", 16)
```
Comment (0xhaven) [inline brainpool/brainpool.go]: I'm don't think these `N` values are set correctly.

**Comment by maxtaco** [comment]
How about some tests?

See https://github.com/keybase/kbpgp/blob/master/test/files/brainpool256.iced, https://github.com/keybase/kbpgp/blob/master/test/files/brainpool384.iced and https://github.com/keybase/kbpgp/blob/master/test/files/brainpool512.iced

File: brainpool/brainpool.go (hunk 1 above)
Comment (maxtaco) [inline brainpool/brainpool.go]: My reference is here: https://github.com/keybase/kbpgp/blob/master/src/ecc/curves.iced#L149-L179

Comment (0xhaven) [inline brainpool/brainpool.go]: Yes, this matches that, although the [elliptic.CurveParams](https://golang.org/pkg/crypto/elliptic/#CurveParams) struct has no `A` member (it's assumed to be `-3`)

**Comment by 0xhaven** [comment]
@maxtaco I've changed how the r curves are generated (using the t curves as twist), necessary because `"crypto/elliptic"` assumes curves are in short-form Weierstrass with `A = -3`.
I've added basic ECDSA tests and OpenPGP parsing tests for those test keys.

File: brainpool/rcurve.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	z3      *big.Int
+	zinv2   *big.Int
+	zinv3   *big.Int
+}
+
+func newrcurve(twisted elliptic.Curve, gx, gy, z *big.Int) *rcurve {
+	var curve rcurve
+
+	curve.twisted = twisted
+	curve.params = *twisted.Params()
```
Comment (0xhaven) [inline brainpool/rcurve.go]: Should probably set `curve.params.Name` here.

**Comment by maxtaco** [comment]
@jacobhaven, can you also provide an End-to-End sign/verify and decrypt/encrypt test?

**Comment by 0xhaven** [comment]
[packet#Signature.Sign](https://github.com/keybase/go-crypto/blob/master/openpgp/packet/signature.go#L536) doesn't appear to support `PubKeyAlgoECDSA`

``` Go
    sig := &packet.Signature{
        PubKeyAlgo: packet.PubKeyAlgoECDSA,
        Hash:       crypto.SHA256,
    }

    h := sig.Hash.New()
    if _, err := h.Write([]byte("Hello World")); err != nil {
        return err
    }

    if err := sig.Sign(h, priv, nil); err != nil {
        return err // "openpgp: unsupported feature: public key algorithm: 19"
    }
```

**Comment by maxtaco** [comment]
That is a slight bummer. Maybe just for now verify a sig that was signed
elsewhere?

**Comment by maxtaco** [comment]
Or does that implicitly happen in key import?

**Comment by 0xhaven** [comment]
I'm trying a different route via https://godoc.org/github.com/keybase/go-crypto/openpgp#DetachSign but that requires parsing a whole `Entity`, not just a private key.

**Comment by maxtaco** [comment]
BTW, another option is to add support for signing in `signature.go`.  It might not be too much heavy-lifting.

**Comment by 0xhaven** [comment]
This package was in a weird in-between state wrt ECDSA support. I went through and made it a first-class algorithm everywhere.

**Comment by maxtaco** [comment]
This looks outstanding.  Can you add one last test which is to verify a
brainpool sig made with gpg or possibly the JavaScript library?

File: brainpool/rcurve.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+func (curve *rcurve) ScalarMult(x1, y1 *big.Int, scalar []byte) (x, y *big.Int) {
+	tx1, ty1 := curve.toTwisted(x1, y1)
+	return curve.fromTwisted(curve.twisted.ScalarMult(tx1, ty1, scalar))
+}
+
+func (curve *rcurve) ScalarBaseMult(scalar []byte) (x, y *big.Int) {
+	return curve.fromTwisted(curve.twisted.ScalarBaseMult(scalar))
+}
```
Comment (maxtaco) [inline brainpool/rcurve.go:83]: can you add an assertion of the form:

``` go
var _ elliptic.Curve = (*rcurve)(nil)
```

I find such things useful for readability.

**Comment by maxtaco** [comment]
A small thing and a few more test cases...

Also, @oconnor663 or @akalin-keybase do you want to take a pass through the crypto? Nice to have a 3rd set of eyes. 

Looks great overall!

File: brainpool/rcurve.go (hunk 3 above)
Comment (0xhaven) [inline brainpool/rcurve.go:83]: Done.

**Comment by 0xhaven** [comment]
*added test of brainpool signatures generated by GPG*

**Comment by maxtaco** [comment]
I'm a :+1:, @oconnor663 said he'll take a pass too, thanks!

File: brainpool/rcurve.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	return &x, &y
+}
+
+func (curve *rcurve) Params() *elliptic.CurveParams {
+	return curve.params
+}
+
+func (curve *rcurve) IsOnCurve(x, y *big.Int) bool {
+	return curve.twisted.IsOnCurve(curve.toTwisted(x, y))
+}
```
Comment (oconnor663) [inline brainpool/rcurve.go:64]: Is this the part that protects us from ["invalid curve attacks"](https://safecurves.cr.yp.to/twist.html)? I know basically nothing about EC, other than that these scary things exist :p

File: brainpool/brainpool_test.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+	msg := []byte("test")
+	r, s, err := ecdsa.Sign(rand.Reader, priv, msg)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !ecdsa.Verify(&priv.PublicKey, msg, r, s) {
+		t.Fatal("signature didn't verify.")
+	}
```
Comment (oconnor663) [inline brainpool/brainpool_test.go:24]: Do we have any test cases that check that bad signatures fail?

**Comment by oconnor663** [comment]
Just the two comments above that I could think of. Thanks Jacob! At some point I should get you to teach me this stuff :)

File: brainpool/rcurve.go (hunk 4 above)
Comment (0xhaven) [inline brainpool/rcurve.go:64]: Yeah, it's checked when [Unmarshaling a point](https://golang.org/src/crypto/elliptic/elliptic.go#L330) onto a curve

File: brainpool/brainpool_test.go (hunk 5 above)
Comment (0xhaven) [inline brainpool/brainpool_test.go:24]: Any suggestions on how to generate a bad signature that could conceivably succeed?

**Comment by maxtaco** [comment]
I'm going to give a :+1: that overrides @oconnor663's request for test if that test is hard to produce.

File: brainpool/rcurve.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	zinv2   *big.Int
+	zinv3   *big.Int
+}
+
+var (
+	two   = big.NewInt(2)
+	three = big.NewInt(3)
+)
+
+func newrcurve(twisted elliptic.Curve, params *elliptic.CurveParams, z *big.Int) *rcurve {
```
Comment (akalin-keybase) [inline brainpool/rcurve.go:26]: noob question: https://golang.org/pkg/crypto/elliptic/#CurveParams says that the `CurveParams` implementation of `Curve` is non-constant time, and we always pass those to `newrcurve`. Is it not a big deal that these curves don't use constant-time ops?

Comment (0xhaven) [inline brainpool/rcurve.go:26]: For the NIST curves, Go only has constant time implementations for P-224 a P-256.
There may be potential for timing attacks here, but they would also apply to NIST P-384 and P-521.

Comment (maxtaco) [inline brainpool/rcurve.go:26]: Yay NaCl

Comment (akalin-keybase) [inline brainpool/rcurve.go:26]: Got it, thanks!

**Comment by akalin-keybase** [comment]
:+1:

**Comment by ebfe** [comment]
If you copy/paste from github.com/ebfe/brainpool you might copy the tests as well...

**Comment by maxtaco** [comment]
@ebfe what is the license on the code in that repo?  @jacobhaven please make an attribution in the comments. And/or is it possible to import @ebfe's library?

**Comment by ebfe** [comment]
> @ebfe what is the license on the code in that repo?
> Well there isn't any license. Not putting a license/readme in the repo was an
> (now obviously failed) attempt at discouraging people to use it for anything
> serious. The code was never reviewed by anybody. In particular the
> implementation of the r-curves is quite hackish (crypto/elliptic assumes all
> curves are short-form weierstrass with a=-3 but the r-curves are not), and I'm
> a bit surprised that didn't come up in this review.

If it helps I can put a BSD-style license on the code and would be happy if
eventual fixes/improvements made their way back home.

**Comment by akalin-keybase** [comment]
@ebfe the a=-3 fact did come up in this review -- see https://github.com/keybase/go-crypto/pull/18#issuecomment-194295022 .

File: brainpool/brainpool.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
+// Package brainpool implements Brainpool elliptic curves.
+// Implementation of rcurves is borrowed from github.com/ebfe/brainpool
```
Comment (maxtaco) [inline brainpool/brainpool.go]: remove word "borrowed"

**Comment by maxtaco** [comment]
I'm a :+1:. @ebfe thank you for your comments and sharing your code with us.  Sorry for the earlier snub, that wasn't cool on our part.

**Pull request merged**